assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,658 @@
+    // read the loop body
+    parse_token(context, tokens, procedures, &mut loop_body)?;
+
+    // consume the `end` token
+    match tokens.read() {
+        None => Err(AssemblyError::unmatched_repeat(
+            tokens.read_at(repeat_start).expect("no repeat token"),
+        )),
+        Some(token) => match token.parts()[0] {
+            Token::END => token.validate_end(),
+            Token::ELSE => Err(AssemblyError::dangling_else(token)),
+            _ => Err(AssemblyError::unmatched_repeat(
+                tokens.read_at(repeat_start).expect("no repeat token"),
+            )),
+        },
+    }?;
+    tokens.advance();
+
+    nodes.push(Node::Repeat(repeat_start, loop_body));
+
+    Ok(())
+}
+
+// parse exec token into AST nodes.
+fn parse_exec(
+    context: &ParserContext,
+    label: String,
+    tokens: &mut TokenStream,
+    procedures: &ProcMap,
+    nodes: &mut Vec<Node>,
+) -> Result<(), AssemblyError> {
+    tokens.advance();
+
+    if label.contains(MODULE_PATH_DELIM) {
+        let full_proc_name = get_full_imported_proc_name(context, label);
+        nodes.push(Node::Instruction(Instruction::ExecImported(full_proc_name)))
+    } else {
+        let index = procedures.get(&label).unwrap().index;
+        nodes.push(Node::Instruction(Instruction::ExecLocal(index)))
+    }
+
+    Ok(())
+}
+
+// parse call token into AST nodes.
+fn parse_call(
+    context: &ParserContext,
+    label: String,
+    tokens: &mut TokenStream,
+    procedures: &ProcMap,
+    nodes: &mut Vec<Node>,
+) -> Result<(), AssemblyError> {
+    tokens.advance();
+    if label.contains(MODULE_PATH_DELIM) {
+        let full_proc_name = get_full_imported_proc_name(context, label);
+        nodes.push(Node::Instruction(Instruction::CallImported(full_proc_name)))
+    } else {
+        let index = procedures.get(&label).unwrap().index;
+        nodes.push(Node::Instruction(Instruction::CallLocal(index)))
+    }
+
+    Ok(())
+}
+
+// Parses a token from the token stream in a body, which generates a series of AST nodes.
+fn parse_token(
+    context: &ParserContext,
+    tokens: &mut TokenStream,
+    procedures: &ProcMap,
+    nodes: &mut Vec<Node>,
+) -> Result<(), AssemblyError> {
+    if let Some(token) = tokens.read() {
+        match token.parts()[0] {
+            Token::ELSE => token.validate_else()?,
+            Token::IF => {
+                token.validate_if()?;
+                parse_if(context, tokens, procedures, nodes)?
+            }
+            Token::WHILE => parse_while(context, tokens, procedures, nodes)?,
+            Token::REPEAT => parse_repeat(context, tokens, procedures, nodes)?,
+            Token::EXEC => {
+                let label = token.parse_exec()?;
+                parse_exec(context, label, tokens, procedures, nodes)?;
+            }
+            Token::CALL => {
+                let label = token.parse_call()?;
+                parse_call(context, label, tokens, procedures, nodes)?;
+            }
```

Comment:
Is there a reason to treat `exec` and `call` differently from any other instruction except for the fact at `is_control_token()` returns true for them? If not, we can probably write our own version of this function for now which excludes these two.

Then, we can handle their parsing in `parse_op_token()`.

Reply:
hmm - I don't think that's the case. by the time we get to `exec` or `call` instructions, the procedures they are invoking should already be parsed. Otherwise, it should be an error.